### PR TITLE
Java pinned memory pool allocator

### PR DESCRIPTION
**Pull request by jlowe**
Adds a pinned memory pool on the Java side along with a runtime Java config option, ai.rapids.cudf.prefer-pinned, that specifies whether host buffers should use pinned memory if possible. Even if pinned memory buffers are preferred, the pinned memory pool must be initialized first or host buffers will fallback to paged memory.

Host buffers that need paged memory can use the new `allocate` overload with a boolean parameter to request the pinned memory allocator not be used.

### CODE REVIEW

**Comment by jrhemstad** [comment]
@harrism should we add a pinned memory resource to RMM?

**Comment by harrism** [comment]
@jrhemstad yes I think that makes sense. In RMM I like the ID of having a non-pool MR for every memory type, and then we can use any of those MRs as a heap resource for a pool MR, making it flexible. All your memory are belong to RMM. :)

File: java/src/main/java/ai/rapids/cudf/PinnedMemoryPool.java [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+
+    boolean canCombine(MemorySection other) {
+      boolean ret = (other.baseAddress + other.size) == baseAddress ||
+          (baseAddress + size) == other.baseAddress;
+      log.trace("CAN {} COMBINE WITH {} ? {}", this, other, ret);
+      return ret;
+    }
+
+    void combineWith(MemorySection other) {
+      assert canCombine(other);
```
Comment (razajafri) [inline java/src/main/java/ai/rapids/cudf/PinnedMemoryPool.java]: should this just be an if?

File: java/src/main/java/ai/rapids/cudf/PinnedMemoryPool.java [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+      log.trace("COMBINING {} AND {}", this, other);
+      this.baseAddress = Math.min(baseAddress, other.baseAddress);
+      this.size = other.size + this.size;
+      log.trace("COMBINED TO {}\n", this);
+    }
+
+    MemorySection splitOff(long newSize) {
+      assert this.size >= newSize;
```
Comment (razajafri) [inline java/src/main/java/ai/rapids/cudf/PinnedMemoryPool.java]: do we have a need for a splitOff memory section to have the same size as the original? Should we instead only check if ```this.size > newSize```

File: java/src/main/java/ai/rapids/cudf/PinnedMemoryPool.java (hunk 1 above)
Comment (revans2) [reply java/src/main/java/ai/rapids/cudf/PinnedMemoryPool.java]: No the code that calls this method should have called `canCombine` already.  No need to call it twice for an API that is not publicly facing.  This is here as a sanity check that we did our job correctly, but not take the cost of it in production.

**Pull request merged**